packages/block-editor/src/store/reducer.js

```diff
@@ -264,16 +281,34 @@ function withIgnoredBlockChange( reducer ) {
  * @return {Function} Enhanced reducer function.
  */
 const withInnerBlocksRemoveCascade = ( reducer ) => ( state, action ) => {
-	if ( state && action.type === 'REMOVE_BLOCKS' ) {
-		const clientIds = [ ...action.clientIds ];
+	const getAllChildren = ( clientIds ) => {
+		const result = [ ...clientIds ];
```

Comment:
While we're simply inheriting precedent here, I wonder if we could optimize this to only create a shallow clone of the array if the loop below would result in an `Array#push`. Seems more likely most blocks would not need to modify the original set (optimizing for non-nested blocks).

We do similar in many of our reducers for objects:

```js
let result = clientIds;

for ( let i = 0; i < result.length; i++ ) {
	if ( ! state.order[ result[ i ] ] ) {
		continue;
	}

	if ( result === clientIds ) {
		result = [ ...result ];
	}

	result.push( ...state.order[ result[ i ] ] );
}

return result;
```

I kinda doubt it would have much impact, and questionable whether the added "complexity" is worthwhile. That said, considering we already have prior art with `getMutateSafeObject` in this function, we could expand it to cover this use-case.

(Not a blocker)